### ISSUE TITLE
Bad alignment in distributed train

### ISSUE THREAD

**Issue by hadaev8** [NVIDIA/tacotron2#285]
Im training model one single v100 and 8xv100 instances.
This is alignment from single gpu train
https://i.imgur.com/byKRmiN.png
This is from multi
https://i.imgur.com/AsYjTbV.png
11648 is equivalent of 93184 step

Any advises?

**Comment by bfs18** [comment]
https://github.com/pytorch/pytorch/issues/2584
This may be caused by BatchNorm1d does not support syncing statistics across GPUs.

**Comment by hadaev8** [comment]
Did you have experience with dist train?
This repo has its own implementation of dataparallel.
I tried to use the native PyTorch function but got a strange error.

https://discuss.pytorch.org/t/distributeddataparallel-do-not-work-with-custom-function-in-model/61631

For example, apex has the implementation of sync batchnorm, but I'm stuck a bit with it

**Comment by bfs18** [comment]
I haven't used dist train.
If the split batch size on a single GPU is large enough, BatchNorm should work.

**Comment by hadaev8** [comment]
I have batch size same as single gpu train

**Comment by rafaelvalle** [comment]
Are you resuming from the pre-trained model? If you are, try decreasing the learning rate when using distributed.

**Comment by hadaev8** [comment]
No, from scratch.
Also, why dont you use pytorch native dataparallel or destributed dataparallel?

**Comment by rafaelvalle** [comment]
We've used this repo to train Tacotron 2 with batch size 64 and 8 V100s on the LJS dataset without sync batch norm nor using the distributed dataparallel that now comes with pytorch.

**Comment by hadaev8** [comment]
@rafaelvalle did you changed lr for distributed train? Increase/decrease?

**Comment by rafaelvalle** [comment]
We used the same learning rate.

**Comment by rafaelvalle** [comment]
@hadaev8 what solved the issue?

**Comment by hadaev8** [comment]
@hadaev8 I moved to pytorch DDP and made a lot of another modifications, where is till gap between single and multi GPU, but it works.

**Comment by xin71** [comment]
Hi, i am also having issue on bad alignment in multi-GPU training. @rafaelvalle , when you trained with 8GPU, how long did it take for you to see an alignment? Was it 8x faster?  @hadaev8 is it possible to share how did you make that work? Is the repo you on your github? Thanks very much for both of you :)

**Comment by hadaev8** [comment]
Try this
https://github.com/NVIDIA/DeepLearningExamples/tree/master/PyTorch/SpeechSynthesis/Tacotron2

**Comment by xin71** [comment]
Thanks for your help! I do see the alignment after changing the learning rate to `1e-3`  which is the one that uses for single-gpu training.

**Comment by xin71** [comment]
@rafaelvalle I am wondering how fast did your model generate a decent alginemnt with 8-gpus training? Was it 8x faster compared with single gpu? Mine takes 26500 steps with 4gpus. (it is equal to about 100, 000 steps if you count with single GPU), but the single GPU only needs about 30, 000 steps. And, I trimed the silence of all the samples so it might look a bit different from yours. Thanks for your time! 

![image](https://user-images.githubusercontent.com/19635856/72836501-9d6e1700-3c41-11ea-8900-005e81054f86.png)

**Comment by hadaev8** [comment]
For me 8GPU set give only x2 time gain in terms of loss.
https://images-ext-2.discordapp.net/external/zOYwL7PShWg9SflWYakeRp4WEin36PujZR6fRkEKwko/https/i.imgur.com/ldW0uQw.png

**Comment by liyang619** [comment]
@xin71  how much epoch equal to 100, 000 steps ?

**Comment by v-nhandt21** [comment]
> [pytorch/pytorch#2584](https://github.com/pytorch/pytorch/issues/2584) This may be caused by BatchNorm1d does not support syncing statistics across GPUs.

I reach the same problem when try to use pytorch DDP, 

Do you remember what you have modified? Can we replace BatchNorm1d with BatchNorm2d to solve the gap?
